### PR TITLE
[DUOS-783][risk=no] Better validation on dataset required fields

### PR DESCRIPTION
**Pull request by lindojong**
Addresses follow up comments on initial PR for DUOS-724 as described here: https://broadinstitute.atlassian.net/browse/DUOS-783

Added validation for dataset name (can't be a duplicate of preexisting name) and # of participants (must be a positive integer). Also replaced DAC text input field with a Select picker. Spoke with Jonathan and confirmed that at this time we still don't need to validate specific urls or vocabulary.

Screen Shot 2020-10-06 at 4 40 18 PM
Screen Shot 2020-10-06 at 4 41 04 PM

Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] PR is labeled with a Jira ticket number and includes a link to the ticket
- [x] PR is labeled with a security risk modifier [no, low, medium, high] 
- [x] PR describes scope of changes

In all cases:

- [ ] Get a minimum of one thumbs worth of review, preferably 2 if enough team members are available
- [ ] Get PO sign-off for all non-trivial UI or workflow changes
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/pages/DatasetRegistration.js [hunk 1]
```diff
@@ -105,6 +112,19 @@ class DatasetRegistration extends Component {
     });
   };
 
+  handlePositiveIntegerOnly = (e) => {
+    const field = e.target.name;
+    const value = e.target.value.replace(/[^\d]/,'');
+
+    if (value === '' || parseInt(value) > -1) {
```
Comment (rushtong) [inline src/pages/DatasetRegistration.js]: Codacy suggestion:
```suggestion
    if (value === '' || parseInt(value, 10) > -1) {
```

File: src/pages/DatasetRegistration.js [hunk 2]
```diff
@@ -126,6 +147,24 @@ class DatasetRegistration extends Component {
       this.isValid(formData.description);
   };
 
+  validateDatasetName(name) {
+    let datasets = this.state.allDatasets;
+    let val = fp.filter(function(ds) {
+      let properties = ds.properties;
+      return fp.find(function(p) { return p.propertyName === "Dataset Name" && p.propertyValue === name}, properties);
```
Comment (rushtong) [inline src/pages/DatasetRegistration.js]: Codacy suggestion:
```suggestion
      return fp.find(function(p) { return p.propertyName === "Dataset Name" && p.propertyValue === name; }, properties);
```

File: src/pages/DatasetRegistration.js [hunk 3]
```diff
@@ -427,16 +489,15 @@ class DatasetRegistration extends Component {
...
                           required: true,
                         }),
                         span({
                           className: 'cancel-color required-field-error-span',
-                          isRendered: fp.isEmpty(this.state.datasetData.datasetName) && showValidationMessages,
+                          isRendered: (fp.isEmpty(this.state.datasetData.datasetName) && showValidationMessages)
+                            || (!fp.isEmpty(this.state.datasetData.datasetName) && !this.validateDatasetName(this.state.datasetData.datasetName)),
                         },
-                        ['Required field']),
+                        [this.validateDatasetName(this.state.datasetData.datasetName) ? 'Required field' : 'Dataset Name already in use']),
```
Comment (rushtong) [inline src/pages/DatasetRegistration.js:522]: I'm not seeing this error pop up. If I create a dataset, then immediately try to create a new one with the same name, I get a 404 back from consent and a generic error message. We might need a bug fix on the POST endpoint ... have not looked closely at that yet.

Screen Shot 2020-10-07 at 2 48 12 PM

Comment (rushtong) [reply src/pages/DatasetRegistration.js:522]: Ancillary problem (and unrelated to this PR) with the above screen shot. Both buttons are disabled and the only way to close out of the modal is to click the little "X" icon.

File: src/pages/DatasetRegistration.js [hunk 4]
```diff
@@ -637,18 +700,25 @@ class DatasetRegistration extends Component {
...
+                          isClearable: true,
+                          isMulti: false,
+                          isSearchable: true,
+                          options: this.dacOptions(),
+                          placeholder: 'Select a DAC...',
                           className: (fp.isEmpty(this.state.datasetData.dac) && showValidationMessages) ?
-                            'form-control required-field-error' :
-                            'form-control',
+                            'required-field-error' :
+                            '',
```
Comment (rushtong) [inline src/pages/DatasetRegistration.js:741]: The required field error label doesn't clear state after being invalid, then selecting a valid one.

Screen Shot 2020-10-07 at 2 52 08 PM

**Review by rushtong — CHANGES_REQUESTED** [review]
Looking good! Still some functional issues here though. See comments inline.

File: src/pages/DatasetRegistration.js (hunk 3 above)
Comment (rushtong) [reply src/pages/DatasetRegistration.js:522]: More info on this. On the back end, we throw a 404 in this case. I think that should really be a 409 Conflict instead. We should add a PR for that separate from this story.

Comment (lindojong) [reply src/pages/DatasetRegistration.js:522]: Made a ticket for the error messaging here: https://broadinstitute.atlassian.net/browse/DUOS-797

Comment (lindojong) [reply src/pages/DatasetRegistration.js:522]: Ticket for modal: https://broadinstitute.atlassian.net/browse/DUOS-800

**Review by rushtong — APPROVED** [review]
Thank you for the changes! 👍

**Pull request merged**